[LTO][AllocToken] Support AllocToken instrumentation in backend

melver opened the pull request:
Unconditionally add AllocTokenPass to the optimization pipelines, and
ensure that it runs last in LTO backend pipelines. The latter ensures
that AllocToken instrumentation can be moved later in the LTO pipeline
to avoid interference with other optimizations (e.g. PGHO) and enable
late heap-allocation optimizations.

In preparation of removing AllocTokenPass being added by Clang, add
support for AllocTokenPass to read configuration options from LLVM
module flags.

To optimize given the pass is now runs unconditionally, only retrieve
TargetLibraryInfo and OptimizationRemarkEmitter when necessary.

---

This change is part of the following series:
1. https://github.com/llvm/llvm-project/pull/169242
2. https://github.com/llvm/llvm-project/pull/169243
3. https://github.com/llvm/llvm-project/pull/169358
4. https://github.com/llvm/llvm-project/pull/169360

melver committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.8-beta.1

melver committed:
[𝘀𝗽𝗿] changes to main this commit is based on

Created using spr 1.3.8-beta.1

[skip ci]

melver committed:
rebase

Created using spr 1.3.8-beta.1

melver committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.8-beta.1

[skip ci]

melver committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.8-beta.1

[skip ci]

melver committed:
rebase

Created using spr 1.3.8-beta.1

melver committed:
rework

Created using spr 1.3.8-beta.1

teresajohnson commented on llvm/lib/Passes/PassBuilderPipelines.cpp:
```diff
@@ -1615,6 +1617,9 @@ PassBuilder::buildModuleOptimizationPipeline(OptimizationLevel Level,
   MPM.addPass(createModuleToFunctionPassAdaptor(std::move(OptimizePM),
                                                 PTO.EagerlyInvalidateAnalyses));
 
+  if (LTOPostLink)
```
Maybe add a comment that for non-LTO compiles this is added to the pipeline by Clang? I suppose you could just enable it here always and remove it from the Clang setup, up to you.

Also, maybe add a comment here that it is done late so that other call transformations, e.g. those done by InstCombine, are complete.

ditto for the main invocation in the full LTO default pipeline.

teresajohnson reviewed (commented):
Could you also add the alloc token metadata, and/or whatever else is needed to cause the allocator calls to kick in, to at least one of the operator new calls in llvm/test/Transforms/InstCombine/simplify-libcalls-new.ll (or create a new small test that has this), and invoke various pipelines on it via opt to ensure we still get the expected transformations (to hot cold operator new and subsequently the alloc token version)? E.g. `opt -passes='thinlto<O2>'` and `opt -passes='lto<O2>'` ?

melver committed:
address review comments

Created using spr 1.3.8-beta.1

melver committed:
simplify copts, add extra tests

Created using spr 1.3.8-beta.1

melver commented:
> > Could you also add the alloc token metadata, and/or whatever else is needed to cause the allocator calls to kick in, to at least one of the operator new calls in llvm/test/Transforms/InstCombine/simplify-libcalls-new.ll (or create a new small test that has this), and invoke various pipelines on it via opt to ensure we still get the expected transformations (to hot cold operator new and subsequently the alloc token version)? E.g. `opt -passes='thinlto<O2>'` and `opt -passes='lto<O2>'` ?
> 
> Largely lgtm to me, except I had a comment about simplifying the cl::opt checks.

Done.

> Also, please add a change to an existing memprof test like suggested above - this will ensure the phase ordering stays correct.

Sorry, missed that.
Added 2 new tests - it seems that we need -supports-hot-cold-new and the index so the LTO pipeline doesn't remove the hot/cold hints, correct? I added a dedicated test to the AllocToken suite and another in the LTO/X86 suite, the latter of which should ensure we catch any deeper regressions.

PTAL.

teresajohnson reviewed (approved):
lgtm

melver committed:
rebase

Created using spr 1.3.8-beta.1
